src/main/java/com/google/sps/servlets/InfoCardServlet.java

```diff
@@ -55,115 +50,30 @@
-
-    try {
-      JSONArray contestListFromApi = ServletUtils.readFromApiUrl(url).getJSONArray("contests");
-      populateDatastoreWithContestEntities(datastore, contestListFromApi);
-      ArrayList<String> contestKeyList = getEntityKeyNameList(datastore, "Contest");
-      chosenElection.setProperty("contests", contestKeyList);
-    } catch (JSONException e) {
-      chosenElection.setProperty("contests", new ArrayList<String>());
-    }
-  }
-
-  /**
-   * Puts candidate Contest Entities in the Datastore.
-   *
-   * @param datastore the Datastore containing all election data
-   * @param contestData API-returned list of contests on the election ballot
-   */
-  private void populateDatastoreWithContestEntities(
-      DatastoreService datastore, JSONArray contestData) {
-    for (Object contestObject : contestData) {
-      JSONObject contest = (JSONObject) contestObject;
-
-      populateDatastoreWithCandidateEntities(datastore, contest.getJSONArray("candidates"));
-
-      Entity contestEntity = Contest.fromJSONObject(contest).toEntity();
-      contestEntity.setProperty("candidates", getEntityKeyNameList(datastore, "Candidate"));
-
-      datastore.put(contestEntity);
-    }
-  }
-
-  /**
-   * Puts Candidate Entities in the Datastore.
-   *
-   * @param datastore the Datastore containing all election data
-   * @param candidateData API-returned list of candidates for a contest on the election ballot
-   * @return list of Key names of all Candidate Entities added
-   */
-  private void populateDatastoreWithCandidateEntities(
-      DatastoreService datastore, JSONArray candidateData) {
-    for (Object candidateObject : candidateData) {
-      JSONObject candidate = (JSONObject) candidateObject;
-      Entity candidateEntity = Candidate.fromJSONObject(candidate).toEntity();
-      datastore.put(candidateEntity);
-    }
-  }
-
-  /**
-   * Gets the names of all Keys of a type of Entity in the Datastore.
-   *
-   * @param datastore the Datastore containing all election data
-   * @param entityType the kind of Entity to query
-   * @return list of Keys' names of all Contest Entities added
-   */
-  public ArrayList<String> getEntityKeyNameList(DatastoreService datastore, String entityType) {
-    ArrayList<String> keyNameList = new ArrayList<String>();
-    Query query = new Query(entityType);
-    PreparedQuery results = datastore.prepare(query);
-
-    for (Entity entity : results.asIterable()) {
-      keyNameList.add(entity.getKey().getName());
-    }
-
-    return keyNameList;
+    String url =
+        String.format(
+            BASE_URL,
+            optionalAddress.get(),
+            optionalElectionId.get(),
+            ServletUtils.getApiKey(PROJECT_ID, SECRET_MANAGER_ID, VERSION_ID));
+    JSONObject voterInfoData = ServletUtils.readFromApiUrl(url);
+    Election election = optionalElection.get();
+    Election.fromVoterInfoQuery(election, datastore, voterInfoData).putInDatastore(datastore);
```

Comment:
I think this will actually insert a duplicate Election entity instead of updating the existing one right?  Can we update instead of creating new Election entities instead?